Rethink Forms.

mpcomplete opened the pull request:
FormField is now a widget that can contain any type of field. Input no
longer has special code to handle form fields. Instead, there is a
helper widget InputFormField for using an Input inside a FormField.

Fixes https://github.com/flutter/flutter/issues/6097 and based on
feedback from the same.

mpcomplete commented:
One more attempt at making sense of Forms. This is based on customer feedback from https://github.com/flutter/flutter/issues/6097 .

There's a number of advantages to this approach, IMO:
1. It works with any type of widget.
2. It's easy to make helper functions/widgets a la InputFormField to simplify creating FormFields that contain a specific widget.
3. FormFields work outside of a Form, if you just want to use it for managing the state of a widget.
4. Forms save and validate as a unit, making it easy to reset the form state (see the ExpansionPanelDemo's cancel).
5. Input no longer has a bunch of special case code for working inside a Form.

Hixie commented on examples/flutter_gallery/lib/demo/expansion_panels_demo.dart:
```diff
@@ -284,22 +291,27 @@ class _ExpansionPanelsDemoState extends State<ExpasionPanelsDemo> {
+                  );
+                },
+              ),
+              onSave: () { form3Key.currentState.save(); close(); },
```
it's sad we have to get the state via a key here... not sure what to suggest instead though

abarth replied:
Maybe onSave should give you a BuildContext you can use to find the Form.

mpcomplete replied:
That could work here where we have a custom widget that implements the onSave function. But sadly there doesn't seem to be a general solution to this issue (see text_field_demo, where we want to call save when a generic RaisedButton is pressed).

I've actually found myself wanting something like this multiple times in this patch - a way for a child widget to refer to its ancestor. The whole reason FormField takes a builder function instead of a child widget is that I want to use FormField's state in the child's builder. A builder function is the only way for child widgets to depend on their ancestor's state in some way.

Hixie commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -116,3 +118,90 @@ class FormScope extends InheritedWidget {
+
+  @override
+  void deactivate() {
+    super.deactivate();
```
nit: call super.deactivate last

mpcomplete replied:
BTW, is there a better way to do this registration? I noticed you did something fancy with inheritFromWidgetOfExactType, but I didn't quite understand it and wasn't sure if it applied here.

Hixie replied:
In my patch the Form didn't need to do anything except cause the registered widgets to rebuild, so it could be vastly simpler. Here you're actually doing things with the widgets, so it's much more complicated.

Hixie commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -116,3 +118,90 @@ class FormScope extends InheritedWidget {
+    super.deactivate();
+    FormScope formScope = FormScope.of(context);
+    if (formScope != null)
+      formScope._formState._unregister(this);
```
rather than these three lines, we should just make the API be `Form.register(context, this);`

Hixie replied:
I guess it doesn't matter since this is all private stuff.

Hixie replied:
This won't actually work. Consider what happens if you're moved around in the tree.

Hixie replied:
I think the right way to do this is to use the pattern I had in the other patch, where Form itself is Inherited and so on, and then for your functions that call things on all the dependents, just add a visitDependents method to InheritedElement that lets you walk all the _dependents, similar to visitChildren, and then call them that way, similar to how I called `dispatchDependenciesChanged` in my `fieldChanged` function in my patch.

mpcomplete replied:
> This won't actually work. Consider what happens if you're moved around in the tree.

I thought `deactivate` was called when this object was moved around in the tree. That's why I'm unregistering here. My understanding was that it would re-register at its new location in the build function. Is this wrong?

Hixie replied:
You're right, my bad.

Hixie replied:
Make sure you have a test for this particular case. I'm not confident we get the order of operations right here. It's possible we detach you from the tree before calling deactivate.

Hixie commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -28,18 +28,27 @@ class Form extends StatefulWidget {
   int generation = 0;
+  Set<FormFieldState> _fields = new Set<FormFieldState>();
 
   void _onFieldChanged() {
```
Our convention is:
* onFoo for variables that take callbacks
* handleFoo for functions that are assigned to onFoo variables
* FooCallback for the types of those variables
* didFoo for methods that you call to notify the object

So I think here you want fieldDidChange or some such.

Hixie commented:
I like this general approach.

Hixie commented:
My main concern is over the need to use GlobalKeys so much to make use of this.

Hixie commented on examples/flutter_gallery/lib/demo/expansion_panels_demo.dart:
```diff
@@ -185,27 +189,30 @@ class _ExpansionPanelsDemoState extends State<ExpasionPanelsDemo> {
           void close() {
             setState(() {
+              // Clear the input state.
+              // TODO(mpcomplete): is there a way to have a Form.reset that purges its children?
```
shouldn't this be straight-forward? I don't really understand the problem here.

mpcomplete replied:
Yes it should! I don't know how, though. :)

Tips?

Hixie commented on examples/flutter_gallery/lib/demo/expansion_panels_demo.dart:
```diff
@@ -185,27 +189,30 @@ class _ExpansionPanelsDemoState extends State<ExpasionPanelsDemo> {
-            onCancel: close
+              onSave: () { form1Key.currentState.save(); close(); },
+              onCancel: close
+            )
```
btw, minor side-note, please move the onSave and onCancel arguments to above the child argument, so that it's clearer which widget it belongs to. Both Adam and I independently misread this the first time (I thought they were on Form, he thought they were on InputFormField).

Hixie commented:
This seems like a net improvement on what we have now, so, modulo the comments above, travis being angry, and the need for tests, LGTM.

We should loop back with customer:leafy once this is in and see how they like it.

mpcomplete committed:
Rethink Forms.

FormField is now a widget that can contain any type of field. Input no
longer has special code to handle form fields. Instead, there is a
helper widget InputFormField for using an Input inside a FormField.

Fixes https://github.com/flutter/flutter/issues/6097 and based on
feedback from the same.

mpcomplete commented:
OK, added a test for the FormState.deactivate case. I think this is ready for commit (pending travis approval).

Hixie commented on examples/flutter_gallery/lib/demo/expansion_panels_demo.dart:
```diff
@@ -189,23 +193,22 @@ class _ExpansionPanelsDemoState extends State<ExpasionPanelsDemo> {
+            key: form1Key,
+            child: new CollapsibleBody(
+              margin: const EdgeInsets.symmetric(horizontal: 16.0),
+              onSave: () { form1Key.currentState.save(); close(); },
```
you're in the form here, so you could avoid using a Key if you instead put a Builder below the form and used Form.of

mpcomplete replied:
Done.

Hixie commented on examples/flutter_gallery/lib/demo/expansion_panels_demo.dart:
```diff
@@ -221,54 +224,57 @@ class _ExpansionPanelsDemoState extends State<ExpasionPanelsDemo> {
+            child: new CollapsibleBody(
+              onSave: () { form2Key.currentState.save(); close(); },
+              onCancel: () { form2Key.currentState.reset(); close(); },
+              child: new FormField<_Location>(
```
are you imagining that eventually we'll have a RadioButtonGroupField or something which generates all the radio buttons and so forth, similar to how we do InputFormField for an Input?

Hixie commented on examples/flutter_gallery/lib/demo/text_field_demo.dart:
```diff
@@ -30,37 +30,39 @@ class TextFieldDemoState extends State<TextFieldDemo> {
-        _validatePhoneNumber(person.phoneNumber) != null ||
-        _validatePassword(person.password) != null) {
+    FormState form = _formKey.currentState;
+    if (form.hasErrors) {
```
nice

Hixie commented on examples/flutter_gallery/lib/demo/text_field_demo.dart:
```diff
@@ -73,55 +75,55 @@ class TextFieldDemoState extends State<TextFieldDemo> {
-                setter:  (String val) { person.name = val; },
-                validator: _validateName
-              )
+            new FormField<InputValue>(
```
why not an InputFormField here?

mpcomplete replied:
I added a comment - I want to demonstrate the equivalence of the 2 ways of doing it.

mpcomplete committed:
comment

Hixie commented:
LGTM modulo minor comments

mpcomplete committed:
form.of
